### PR TITLE
Value type hint as string

### LINKED ISSUE #53424: Parse user type hints directly to ValuesSourceTypes
In the current implementation, user type hints (specified via the `value_type` field on aggregations) get parsed leniently through the `ValueType` enum.  The `ValueType` enum doesn't really add anything however, and we immediately turn that input into a `ValuesSourceType`.  Ideally, we should just parse the user's input directly into a `ValuesSourceType`

While just getting parity with the current parsing would be sufficient, ideally we would tie into the `ValuesSourceRegistry` to allow for users hinting VSTs from plugins.

Relates to #42949

### PR DESCRIPTION
**Pull request by not-napoleon**
Resolves #53424

This stores user value type hints internally as strings, converting them to `ValuesSourceTypes` at config resolution time.

### CODE REVIEW

File: server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/bucket/terms/StringTermsIT.java [hunk 1]
```diff
@@ -1389,16 +1388,20 @@ public void testScriptWithValueType() throws Exception {
             assertThat(terms.getBuckets().size(), equalTo(1));
         }
 
-        String invalidValueType = source.replaceAll("\"value_type\":\"n.*\"", "\"value_type\":\"foobar\"");
+        String invalidValueType = source.replaceAll("\"value_type\":\"n.*?\"", "\"value_type\":\"foobar\"");
```
Comment (not-napoleon) [inline server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/bucket/terms/StringTermsIT.java:1391]: Regexes are hard.  This test has always produced invalid json here, but apparently in the old code path we failed on value type validation first.  Stream parsing I guess.

File: server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/bucket/terms/StringTermsIT.java [hunk 2]
```diff
@@ -1389,16 +1388,20 @@ public void testScriptWithValueType() throws Exception {
...
 
         try (XContentParser parser = createParser(JsonXContent.jsonXContent, invalidValueType)) {
-            XContentParseException ex = expectThrows(
-                XContentParseException.class,
+            SearchPhaseExecutionException ex = expectThrows(
+                SearchPhaseExecutionException.class,
```
Comment (not-napoleon) [inline server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/bucket/terms/StringTermsIT.java:1395]: We no longer catch invalid value types at parse time.  We do still catch them before aggregation execution though (at ValuesSourceConfig#Resolve time).

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java [hunk 3]
```diff
@@ -151,13 +181,17 @@ public ValuesSourceRegistry build() {
...
 
     public ValuesSourceRegistry(
         Map<RegistryKey<?>, List<Map.Entry<ValuesSourceType, ?>>> aggregatorRegistry,
-        AggregationUsageService usageService
+        AggregationUsageService usageService,
+        Map<String, ValuesSourceType> valueTypeLookup
     ) {
         this.aggregatorRegistry = copyMap(aggregatorRegistry);
         this.usageService = usageService;
+        // TODO: Make an immutable copy blah blah blah
```
Comment (imotov) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java:193]: Is this planned for a follow-up?

**Review by imotov — COMMENTED** [review]
The change look reasonable to me. Since it is pretty big, I might have missed something. So, it might make sense for @nik9000 to have another look as well.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java (hunk 3 above)
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java:193]: yeah, I skipped it for time when doing the initial draft, but I'll add that in before merging.  Thanks for the reminder.  Probably a few other small clean ups I need to do too (like fixing the merge conflicts...).

File: x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregatorMetricsTests.java [hunk 4]
```diff
@@ -203,7 +207,8 @@ private ValuesSourceConfig toConfig(SortedNumericDocValues values) throws IOExce
         when(source.isFloatingPoint()).thenReturn(false);
         when(source.longValues(null)).thenReturn(values);
         if (randomBoolean()) {
-            return toConfig(source, CoreValuesSourceType.NUMERIC, randomWholeNumberDocValuesFormat(), true);
+            // TODO: should we have a case for CVST.LONG here?
+            return toConfig(source, CoreValuesSourceType.DOUBLE, randomWholeNumberDocValuesFormat(), true);
```
Comment (nik9000) [inline x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregatorMetricsTests.java:211]: It's probably worth it now that they are separate types. But it's not a blocker for this at all.

**Review by nik9000 — CHANGES_REQUESTED** [review]
I think there are two outwards facing changes.

1. Aggs can declare different support for `long` vs `double`. I wonder if we can drop `isFloatingPoint` in a follow up change and have aggs declare support for `long` and `double` separately instead of checking that.
2. Plugins can now declare new values source types. Aggs won't support them out of the box, but the plugin could declare those too. That's neat.


I think I found a `NOCOMMIT` that snuck in. I left a few small requests otherwise.

**Comment by not-napoleon** [comment]
there's no path to merging this at this point.

**Pull request closed without merging**